js/components/operatorSignIn/expiry.tsx

```diff
@@ -0,0 +1,163 @@
+    return "expired today";
+  } else if (delta <= 1) {
+    return "expires tomorrow";
+  } else {
+    return `expires in ${wholeAbs} days`;
+  }
+};
+
+const CertificateBox = ({
+  now,
+  mode,
+  title,
+  operatorName,
+  certifications,
+}: {
+  now: DateTime;
+  mode: "warning" | "error";
+  title: string;
+  operatorName: string;
+  certifications: Certification[];
+}): ReactElement => {
+  const innerString =
+    `Our records show that ${operatorName}'s ` +
+    certifications
+      .map(
+        (c) =>
+          `${humanReadableType(c.type)} ${englishDaysBetween(
+            now,
+            c.expires,
+          )} on ${c.expires.toLocaleString(DateTime.DATE_SHORT)}`,
+      )
+      .join(" and ") +
+    ".";
+  return (
+    <WarningParagraph
+      className={className([
+        "border-0 light:bg-opacity-40 dark:bg-opacity-30 dark:text-white",
+        mode === "warning" && "bg-[#FFDE9E]",
+        mode === "error" && "bg-[#FF919A]",
+      ])}
+    >
+      <p className="font-bold uppercase">{title}</p>
+      <p className="mt-2">{innerString}</p>
+      {mode === "warning" && (
+        <p className="mt-2">Please have them call the Office.</p>
+      )}
+    </WarningParagraph>
+  );
+};
+
+export const CertificateBoxes = ({
+  now,
+  displayName,
+  ignoreExpired,
+  certifications,
+}: {
+  now: DateTime;
+  displayName: string;
+  ignoreExpired: boolean;
+  certifications: Certification[];
+}): ReactElement => {
+  const expired = certifications.filter((cert) => isExpired(cert, now));
+  const expiresSoon = certifications.filter((cert) => {
+    const delta = daysBetween(now, cert.expires);
+    return !isExpired(cert, now) && delta <= WARN_WITHIN_D;
+  });
+  return (
+    <>
+      {expiresSoon.length > 0 && (
+        <CertificateBox
+          now={now}
+          mode="warning"
+          title={
+            expiresSoon.length === 2 ? "Cards expire soon" : "Card expires soon"
+          }
+          operatorName={displayName}
+          certifications={expiresSoon}
+        />
+      )}
+      {expired.length > 0 && !ignoreExpired && (
+        <CertificateBox
+          now={now}
+          mode="error"
+          title={expired.length === 2 ? "Expired cards" : "Expired card"}
+          operatorName={displayName}
+          certifications={expired}
+        />
+      )}
+    </>
+  );
+};
+
+export const Instructions = ({
+  displayName,
+}: {
+  displayName: string;
+}): ReactElement => {
+  return (
+    <ol className="m-8 mr-0 list-decimal">
+      <li>Do not allow {displayName} to drive.</li>
+      <li>Call the Office.</li>
+      <li>Send {displayName} to the Supervisors&#39; Office.</li>
+    </ol>
+  );
+};
+export const Bypass = ({
+  displayName,
+  certifications,
+  now,
+  onContinue,
+}: {
+  displayName: string;
+  certifications: Certification[];
+  now: DateTime;
+  onContinue: () => void;
+}): ReactElement => {
+  const expireds = certifications.filter((c) => isExpired(c, now));
```

Comment:
I don't think this would actually cause problems, but it seems like we shouldn't have to recalculate which cards are expired at this point. The calculations are all in sync now but what if something changes? Perhaps the caller should calculate `expireds` once, and use the same value both for deciding whether to show `Bypass` and for passing in as an argument:

```tsx
expireds = certifications.filter((c) => isExpired(c, now));
...
(expireds.length > 0) ? <Bypass expireds={expireds} displayName={...} onContinue={...} />
```

Reply:
fixed 👍 thanks